### PR TITLE
fix(security): block SQLi via chart_data.where in slimstat_fetch_chart_data

### PR DESCRIPTION
**Pull request by parhumm**
## Summary

Patches a Patchstack-disclosed authenticated SQL Injection (CVSS 8.5, High) in the `slimstat_fetch_chart_data` AJAX endpoint. The `chart_data.where` JSON parameter was concatenated directly into the SQL query passed to `Query::whereRaw()` with no parameter binding. Any logged-in **Subscriber** who could obtain the chart nonce — exposed on any frontend page that renders a Slimstat chart shortcode like `[slimstat f="widget" w="slim_p1_01"]` — could execute time-based blind SQL injection via a payload that wraps `SLEEP()` in a conditional `IF(...)`.

The `data1` / `data2` allowlist introduced in PR #232 already covered the aggregate-expression vector; this PR closes the parallel `where` vector.

**Patchstack disclosure date:** 18 May 2026. This needs to ride 5.4.12.

## Approach

Validate the supplied `where` clause against an allowlist built at request time from `wp_slimstat_reports::$reports`. Every WHERE clause declared by a registered report — including any added by Pro addons via the `slimstat_reports_info` filter — is harvested, whitespace-normalized, and compared for equality. Anything not in the allowlist throws `\Exception`, which the handler already catches and returns as `wp_send_json_error()` (same shape as the existing `validateSqlExpression()` failures).

Why allowlist over server-side `report_id` resolution: the chart JS ships the entire `args` blob and has no awareness of the report ID it belongs to. A registry-based allowlist auto-handles third-party reports (their clauses join the allowlist after the filter fires) and dynamic clauses like `slim_p3_01` (interpolates `home_url()`) and `slim_p7_02` (interpolates `date_i18n('U') - 300`) since the allowlist rebuilds per request.

Capability check intentionally stays at `'read'`. Tightening would break the legitimate Subscriber-views-shortcode-on-frontend-page flow that the threat model itself depends on.

## What changed

- **`src/Modules/Chart.php`** — new `getAllowedWhereClauses()` and `normalizeSqlWhitespace()`; vulnerable concat at `sqlFor()` replaced with allowlist check.
- **`tests/e2e/chart-where-allowlist-validation.spec.ts`** — 6-case regression test, including the Patchstack PoC with a sub-1500ms timing assertion that proves the SLEEP payload was rejected before reaching the query layer.
- **`tests/e2e/helpers/chart.ts`** — extracted `getChartNonce()` and `CHART_TEST_RANGE` so both chart specs share them. Existing `chart-sql-expression-validation.spec.ts` refactored to use them.
- **`CHANGELOG.md` / `readme.txt`** — Security entry added to the existing 5.4.12 changelog block. No version bump (5.4.12 is still pending release).

## Test plan

- [ ] `composer test:all` (PHP unit suite — no impact expected)
- [ ] `npx playwright test tests/e2e/chart-where-allowlist-validation.spec.ts` — all 6 cases green, including the timing assertion on the SLEEP payload
- [ ] `npx playwright test tests/e2e/chart-sql-expression-validation.spec.ts` — existing 4 cases still green after helper extraction
- [ ] Manual repro of the Patchstack PoC against the patched build — Subscriber + extracted nonce + crafted POST returns `{"success":false}` in well under the SLEEP delay
- [ ] Smoke-test admin dashboard charts that DO use a registered `where` (Audience → Search Terms `slim_p1_19_01`, Site Analysis → Human Visits `slim_p2_01`) — granularity changes still re-render
- [ ] Smoke-test as Subscriber against a public page with `[slimstat f="widget" w="slim_p2_01"]` — chart still renders/redraws (regression-risk scenario for the allowlist)

## Summary by CodeRabbit

* **Security**
  * Chart data reporting now validates user-supplied WHERE filters against a trusted allowlist of report-declared clauses; changelog and readme updated.

* **Tests**
  * Added end-to-end tests that exercise allowed WHERE clauses, reject SQLi-like payloads, include shared test helpers, and use a stable chart time range for reliable E2E runs.

### CODE REVIEW

**Comment by parhumm** [comment]
## QA — Smoke quick-test checklist + side-effects

The fix changes runtime behavior on **every** call to `wp_ajax_slimstat_fetch_chart_data`, so QA needs to confirm both (a) charts that previously worked still work, and (b) the rejection path actually fires for malformed input. Use the matrix below.

### What the patch does at runtime (the why behind every test)

1. On every chart AJAX call, the handler now harvests `callback_args.chart_data.where` from every entry in `wp_slimstat_reports::$reports` (after `wp_slimstat_reports::init()` runs and the `slimstat_reports_info` filter has fired) and treats that as the allowlist.
2. The submitted `chart_data.where` is whitespace-normalized (`/\s+/` → single space, then trimmed) and the same normalization runs on the registered side. Equality must match exactly.
3. On mismatch: `wp_send_json_error(['message' => 'Invalid chart filter expression.'])` (HTTP 200, `success: false`).
4. The allowlist is cached in a `static $cache` for the lifetime of the PHP request only. No transient, no object-cache.

That means the regression risk is concentrated on **reports that DO use a `where` clause**. Reports without `where` are untouched.

---

### Smoke matrix — happy path (must all pass)

#### Admin dashboard (logged in as Administrator)

- [x] **Real-Time** view (`?page=wp-slim-view-1`) — every widget renders, no JS console errors, no `Invalid chart filter expression` in DevTools Network panel
- [x] **Overview** view (`?page=wp-slim-view-2`)
- [x] **Audience** view (`?page=wp-slim-view-3`) — **important: this includes `slim_p1_19_01` Search Terms which uses a registered `where` clause**
- [x] **Site Analysis** view (`?page=wp-slim-view-4`) — **important: this includes `slim_p2_01` Human Visits which uses a registered `where` clause**
- [x] **Traffic Sources** view (`?page=wp-slim-view-5`)
- [x] On any page with a chart: change **granularity** (Hourly / Daily / Weekly / Monthly / Yearly) → AJAX returns `success: true`, chart redraws
- [x] On any page with a chart: change **date range** (custom range, last 7 days, last 30 days) → AJAX returns `success: true`, chart redraws
- [x] **Customize widgets** (drag, hide/show) → after save, hidden widgets that had a `where` reappear correctly when re-enabled

#### Frontend chart shortcodes

For each, test as **Administrator**, then **Subscriber**, then **anonymous**:

- [x] `[slimstat f="widget" w="slim_p1_01"]` (Pageviews — **no `where`**) — renders
- [x] `[slimstat f="widget" w="slim_p1_19_01"]` (Search Terms — **has `where`**) — renders, granularity dropdown still works
- [x] `[slimstat f="widget" w="slim_p2_01"]` (Human Visits — **has `where`**) — renders, granularity dropdown still works
- [x] `[slimstat f="widget" w="slim_p3_01"]` (uses **dynamic** `where` interpolating `home_url()`) — confirm no false rejection
- [ ] `[slimstat f="widget" w="slim_p7_02"]` (uses **dynamic** `where` interpolating `date_i18n('U') - 300`) — confirm no false rejection

> **Why the dynamic-clause cases matter:** the allowlist is rebuilt per request, so `home_url()` and `date_i18n()` are evaluated fresh on each call. If anything in the harvest path is cached too aggressively (or the request runs against a stale `wp_slimstat_reports::$reports` snapshot), these are the first to break.

### Smoke matrix — rejection path (must reject)

These confirm the fix actually fires. Run as a logged-in user with a captured `slimstat_chart_nonce` (read it from the page source on any frontend page that renders a chart shortcode).

POST to `/wp-admin/admin-ajax.php` with `action=slimstat_fetch_chart_data`, the captured nonce, `granularity=daily`, and `args` JSON:

- [ ] `chart_data.where` containing a `SLEEP()` payload wrapped in `IF(1=1,...,0)` → response under 1.5s, body `{"success":false,"data":{"message":"Invalid chart filter expression."}}`
- [ ] `chart_data.where = "1=1; DROP TABLE wp_slim_stats--"` → `success:false`, table still present afterwards
- [ ] `chart_data.where = "anything totally made up"` → `success:false`
- [ ] **No `where`** field at all → `success:true` (regression check — this is most charts)

The automated regression test `tests/e2e/chart-where-allowlist-validation.spec.ts` covers all four — please confirm it actually runs in your CI pipeline and goes green, not just exists.

---

### Side-effects to watch for (non-obvious)

1. **Pro addon reports registered via `slimstat_reports_info` filter must register at plugin-load time.** If any addon registers its `where`-using report from a *late-firing* hook (e.g. `init` priority > 10, or worse, an admin-only hook), the report's `where` clause will NOT be in the allowlist when the AJAX request fires, and the chart will return `Invalid chart filter expression`. **QA action:** if testing with Pro installed, exercise every Pro chart that uses a `where`. Symptom of breakage: chart fails to load with the rejection message in the AJAX response.

2. **Custom integrations / third-party code that calls `slimstat_fetch_chart_data` directly with a hand-crafted `chart_data.where`** will break. This is intended (security boundary) but may surprise users who built custom dashboards or external scripts hitting this endpoint. **QA action:** ask support team if any known custom integrations exist; document the breaking change in the support knowledge base.

3. **The error message `"Invalid chart filter expression."` is new** and is not yet in any translation file. Non-English sites will see English fallback until the next .pot regeneration + translation pass. Low impact (most users will never see it).

4. **Performance baseline shifts slightly** for the slimstat AJAX endpoint when the request is the FIRST in the PHP process to load the reports registry. That happens for **frontend Subscriber-initiated AJAX**, where `admin/view/wp-slimstat-reports.php` (~2200 lines) hadn't been loaded yet. Cost: one file include + `wp_slimstat_reports::init()` (which short-circuits if already initialized). For admin dashboard AJAX the file is normally already loaded so it's free. **QA action:** if you have the perf k6 suite wired up, run `npm run test:perf` against the chart endpoint and compare p95 latency vs. main — should be within noise.

5. **Whitespace normalization is one-way (collapse + trim).** If a registered `where` clause contains `IS NOT NULL` (with single spaces), and the frontend payload sends `IS  NOT  NULL` (double spaces) due to JSON minification quirks, the allowlist match still works — confirmed by the "extra whitespace" regression test. But if a clause contains semantically-significant repeated whitespace inside a string literal (e.g. `LIKE "% % %"` — a search for spaces), normalization would collapse it. None of the bundled reports do this. Worth flagging if any third-party report does.

6. **No `where` ≠ empty `where`.** Both are accepted (the `if (!empty(...))` guard returns true for both `null`/missing and `""`). This matters because some frontend code paths may serialize an explicit empty string vs. omit the key — both must keep working.

7. **The `static $cache` lives only for the PHP request.** WP-CLI long-running scripts (`wp eval-file`, `wp cron event run` of slimstat crons) have a single request lifetime, so cache freshness is fine. **No** issue with PHP-FPM workers (each request is fresh PHP state from caller's perspective). Confirmed via the efficiency review during /simplify.

8. **`Chart::ajaxFetchChartData()` no longer logs which check failed.** The error message is intentionally generic ("Invalid chart filter expression.") to avoid leaking which clause the registry has. If a legitimate chart starts failing post-merge, the only debug signal is the AJAX response — no PHP error log entry. **QA action:** if a regression is reported, ask the user to: (a) screenshot the failing chart, (b) capture the failing request payload from DevTools Network panel, (c) note the `w=` shortcode parameter or admin page URL.

---

### Quick CI sanity (before merging)

- [ ] `php -l src/Modules/Chart.php` → no syntax errors (already verified locally)
- [ ] `composer test:all` (PHP unit) → all green
- [ ] `npx playwright test tests/e2e/chart-sql-expression-validation.spec.ts` → all 4 green (the existing data1 spec, refactored to share helpers in this PR)
- [ ] `npx playwright test tests/e2e/chart-where-allowlist-validation.spec.ts` → all 6 green (the new spec)

If any chart in section 1 (admin dashboard) or section 2 (frontend shortcodes) breaks: the most likely culprit is whitespace normalization mismatch on a third-party-registered `where` clause. Quickest debug: temporarily `error_log(print_r(self::getAllowedWhereClauses(), true))` at the top of `Chart::sqlFor()` and compare against the failing payload.

**Comment by parhumm** [comment]
## Review response

### ✅ Accepted — `is_string()` guard on AJAX-side `where` input

Fixed in `e1975998`.

Chart.php does not `declare(strict_types=1)`, so the old `(string) $args['chart_data']['where']` cast would:
- emit `E_WARNING` "Array to string conversion" for array input (leaky log, but functionally OK)
- throw a fatal `Error` for an object without `__toString` → HTTP 500 instead of the generic rejection

`is_string()` guard now runs first and routes both cases to the same `Invalid chart filter expression.` response, symmetric to the registry-side guard in `getAllowedWhereClauses()` (applied in `7ecc1423`).

### ❌ Rejected — also harvest top-level `callback_args['where']`

Not needed. Verified against the current code:

**1. The chart AJAX handler only ever reads `$args['chart_data']['where']`** — `Chart::sqlFor()` at the current version line 277, 278, and the block I just hardened. No code path promotes top-level `callback_args['where']` into `chart_data.where`.

**2. The 28 top-level `'where' =>` occurrences in `admin/view/wp-slimstat-reports.php` belong to non-chart reports** (`raw_results_to_html` callbacks). Examples:
- `slim_p1_06` Recent Search Terms (line 162) — renders as HTML table via `wp_slimstat_db::get_recent`
- `slim_p1_22_01` Mobile Devices (line 542) — table via `wp_slimstat_db::get_top`
- `slim_p1_13` User Activity (line 264) — recent-events renderer

These go through completely different server-side code paths and are never shipped to the frontend JS as `chart_data`. The only chart reports with a `where` clause are the three that already declare it under `chart_data`:
- `slim_p1_19_01` Search Terms chart (line 280)
- `slim_p2_01` Human Visits chart (line 300)
- `slim_p3_01` Traffic Sources chart (line 569)

**3. Adding the 28 table-report clauses to the chart allowlist would widen the accepted input set without fixing any code path the chart endpoint actually uses.** While all 28 are author-trusted strings (so the widening is security-neutral), they expand what a Subscriber can submit with no legitimate caller to match. If a Pro addon puts `where` at the wrong level in a chart report, that's a schema bug in the addon — accommodating it silently would hide the mistake.

If a real chart report ever emerges that declares `where` at the top level of `callback_args` (not under `chart_data`), the right fix is to update that report's schema, not to widen the allowlist here.

**Comment by parhumm** [comment]
## Live PoC verification against patched build

Reporter (hhhai) said the PoC still works. I reproduced their exact payload against this branch's HEAD (`e1975998`) on a local 5.4.12 install to verify.

### Setup

- Plugin version loaded by WP: **5.4.12** (patched branch)
- `Chart` class file: `wp-content/plugins/wp-slimstat/src/Modules/Chart.php` mtime `2026-04-26T12:35:15Z`
- Subscriber session, `current_user_can('read')` → YES
- Fresh `slimstat_chart_nonce` minted via `wp_create_nonce()`
- Payload: byte-identical to the reporter's `args` JSON, including the appended `=0` to make it a valid boolean expression:
  ```
  chart_data.where = "IF(1=1,SLEEP(1),0)=0"
  ```

### Result — malicious payload

```json
{"success":false,"data":{"message":"Invalid chart filter expression."}}
```

**Elapsed: 7.89 ms.** SLEEP(1) on this MySQL would take ~10 s if the query reached the database. It didn't — the allowlist gate at [src/Modules/Chart.php:300-315](src/Modules/Chart.php#L300-L315) rejected before query construction. Verdict: **PATCH WORKING ✓**.

### Result — legitimate allowlisted clause (regression check)

Same harness, with `chart_data.where = '(visit_id > 0 AND browser_type <> 1)'` (the registered `slim_p2_01` Human Visits clause):

```json
{"success":true,"data":{"args":{...},"data":{"labels":[...]},"totals":...}}
```

Elapsed: 132.98 ms. Chart data returned normally. The legit path still works.

### Allowlist contents at request time

`Chart::getAllowedWhereClauses()` returned 3 clauses, harvested from `wp_slimstat_reports::$reports`:

1. `searchterms <> "_" AND searchterms IS NOT NULL AND searchterms <> ""` (`slim_p1_19_01`)
2. `(visit_id > 0 AND browser_type <> 1)` (`slim_p2_01`)
3. `(referer IS NOT NULL AND referer NOT LIKE "%http://localhost:10003%")` (`slim_p3_01`, dynamic — `home_url()` interpolated at request time)

The reporter's payload `IF(1=1,SLEEP(1),0)=0` is not byte-equal to any of these after whitespace normalization, so `isset()` returns false and the handler throws.

### Why the reporter still sees the vulnerability

They're almost certainly testing against **wp.org-published 5.4.11**, which is the current Stable tag. This branch's fix is staged as 5.4.12 but **has not yet been released**. Until 5.4.12 ships:
- Every wp.org auto-update install remains on 5.4.11
- The reporter's PoC succeeds against 5.4.11
- Disclosure date: **2026-05-18** — ~22 days

### Recommendation

**Merge this PR and ship 5.4.12 to wp.org as soon as possible.** The fix is verified working end-to-end (CI E2E tier 2 passed in 15m4s, including the new `chart-where-allowlist-validation` spec which encodes this exact PoC with a `<1500ms` timing assertion). The window between disclosure (2026-05-18) and patch availability is the window of maximum risk for users — every day the release waits is a day Subscribers on production sites can be exploited.

If the reporter has a payload that still works against this branch (not against 5.4.11), please ask them to share the exact `args` JSON + nonce flow — the audit found no alternative vector, but a concrete payload would be the fastest way to verify either way.
